pkg/alertmanager/amcfg.go

```diff
@@ -0,0 +1,499 @@
+
+func loadCfg(s string) (*alertmanagerConfig, error) {
+	// Run upstream Load function to get any validation checks that it runs.
+	_, err := config.Load(s)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := &alertmanagerConfig{}
+	err = yaml.UnmarshalStrict([]byte(s), cfg)
+
+	return cfg, nil
+}
+
+func (c alertmanagerConfig) String() string {
+	b, err := yaml.Marshal(c)
+	if err != nil {
+		return fmt.Sprintf("<error creating config string: %s>", err)
+	}
+	return string(b)
+}
+
+func prefixRouteReceivers(r *monitoringv1.Route, prefix string) {
+	if !strings.HasPrefix(r.Receiver, prefix) {
+		r.Receiver = prefix + r.Receiver
+	}
+
+	if r.Routes != nil {
+		for i := range r.Routes {
+			prefixRouteReceivers(&r.Routes[i], prefix)
+		}
+	}
+}
+
+type configGenerator struct {
+	logger        log.Logger
+	kclient       kubernetes.Interface
+	nsSecretCache map[string]*corev1.Secret
+}
+
+func newConfigGenerator(logger log.Logger, kclient kubernetes.Interface) *configGenerator {
+	cg := &configGenerator{
+		logger:        logger,
+		kclient:       kclient,
+		nsSecretCache: make(map[string]*corev1.Secret),
+	}
+	return cg
+}
+
+func (cg *configGenerator) generateConfig(
+	baseConfig alertmanagerConfig,
+	amConfigs map[string]*monitoringv1.AlertmanagerConfig,
+) ([]byte, error) {
+
+	// amConfigIdentifiers is a sorted slice of keys from
+	// amConfigs map, used to always generate the config in the
+	// same order
+	amConfigIdentifiers := make([]string, len(amConfigs))
+	i := 0
+	for k := range amConfigs {
+		amConfigIdentifiers[i] = k
+		i++
+	}
+	sort.Strings(amConfigIdentifiers)
+
+	subRoutes := []*route{}
+	for _, amConfigIdentifier := range amConfigIdentifiers {
+		crKey := types.NamespacedName{
+			Name:      amConfigs[amConfigIdentifier].Name,
+			Namespace: amConfigs[amConfigIdentifier].Namespace,
+		}
+
+		// add routes to subRoutes
+		subRoutes = append(subRoutes, convertRoute(amConfigs[amConfigIdentifier].Spec.Route, crKey, true))
+
+		// add receivers to baseConfig.Receivers
+		for _, receiver := range amConfigs[amConfigIdentifier].Spec.Receivers {
+			baseConfig.Receivers = append(baseConfig.Receivers, cg.convertReceiver(&receiver, crKey))
+		}
+
+		// add inhibitRules to baseConfig.InhibitRules
+		for _, inhibitRule := range amConfigs[amConfigIdentifier].Spec.InhibitRules {
+			baseConfig.InhibitRules = append(baseConfig.InhibitRules, convertInhibitRule(&inhibitRule))
+		}
+	}
+
+	// Append subroutes from base to the end, then replace with the new slice
+	subRoutes = append(subRoutes, baseConfig.Route.Routes...)
+	baseConfig.Route.Routes = subRoutes
+
+	return yaml.Marshal(baseConfig)
+}
+
+func convertRoute(in *monitoringv1.Route, crKey types.NamespacedName, firstLevelRoute bool) *route {
+
+	// Enforce continue to be true for main Route in a CR
+	cont := in.Continue
+	if firstLevelRoute {
+		cont = true
+	}
+
+	match := map[string]string{}
+	matchRE := map[string]string{}
+	if firstLevelRoute {
+		match["namespace"] = crKey.Namespace
```

Comment:
I'm not 100% sure about this -- should it replace _all_ existing matchers with a single namespace matcher like this, or should it just replace existing _namespace_ matchers with this one?

Reply:
we need to override any existing `namespace` matchers.

Reply:
Done.